src/transformers/benchmark/benchmark.py

```diff
@@ -0,0 +1,133 @@
+
+            loss.backward()
+            model.zero_grad()
+
+        if trace_memory is True:
+            if self.args.trace_memory_line_by_line or self.args.n_gpu == 0:
+                trace = start_memory_tracing("transformers")
+            else:
+                # clear cuda cache
+                torch.cuda.empty_cache()
+                torch.cuda.reset_peak_memory_stats()
+
+            # calculate loss and do backpropagation
+            compute_loss_and_backprob()
+
+            if self.args.trace_memory_line_by_line or self.args.n_gpu == 0:
+                summary = stop_memory_tracing(trace)
+                memory = summary.total
+            else:
+                memory = Memory(torch.cuda.max_memory_reserved())
+
+            return memory
+        else:
+            # as written in https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat, min should be taken rather than the average
+            runtimes = timeit.repeat(lambda: compute_loss_and_backprob(), repeat=self.args.repeat, number=10,)
+            return min(runtimes) / 10.0
+
+    def inference(self, model_name, batch_size, sequence_length, trace_memory=False):
+        config = self.config_dict[model_name]
+        model = MODEL_MAPPING[config.__class__](config)
+        model.to(self.args.device)
+        model.eval()
+
+        input_ids = torch.randint(
+            config.vocab_size, (batch_size, sequence_length), dtype=torch.long, device=self.args.device
+        )
+        if trace_memory is True:
+            if self.args.trace_memory_line_by_line or self.args.n_gpu == 0:
+                trace = start_memory_tracing("transformers")
+            else:
+                # clear cuda cache
+                torch.cuda.empty_cache()
+                if hasattr(torch.cuda, "max_memory_reserved"):
+                    torch.cuda.reset_peak_memory_stats()
+                else:
+                    logger.info("Please consider updating PyTorch to version 1.4 to get more accuracy on GPU memory usage")
+                    torch.cuda.reset_max_memory_cached()
+
+            model(input_ids)
+
+            if self.args.trace_memory_line_by_line or self.args.n_gpu == 0:
+                summary = stop_memory_tracing(trace)
+                memory = summary.total
+            else:
+                if hasattr(torch.cuda, "max_memory_reserved"):
+                    memory = Memory(torch.cuda.max_memory_reserved())
+                else:
+                    logger.info("Please consider updating PyTorch to version 1.4 to get more accuracy on GPU memory usage")
+                    memory = Memory(torch.cuda.max_memory_cached())
+
+            return memory
+        else:
+            # as written in https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat, min should be taken rather than the average
+            runtimes = timeit.repeat(lambda: model(input_ids), repeat=self.args.repeat, number=10,)
+            return min(runtimes) / 10.0
```

Comment:
Might have been there before your refactor, but I believe it's usually better to time the average rather than the minimum. The average takes into account hardware stability, and is less likely to bias the results. 

I would advocate for the more stable approach, knowing how the model will usually perform on a given hardware, rather than the absolute best the model will perform, sampled from 10 runs. Open to criticism though, I had opted for minimum as well in the first version ([original benchmark script before TensorFlow team and François Chollet's comments](https://gist.github.com/LysandreJik/c174a4fdc43004f5a2389f81d11444cd/5362a9236ed1d1739df7caa9e0d83a65ee07f2d0)), before switching back to mean.

Reply:
On a second note, you seem to be running inference on 10 runs right after initializing the model. This isn't problematic in most cases, but would be for cases where the model must be traced before being run (see TorchScript, TPU).

On TPU for example, it's not surprising to have the first iteration run for 45 seconds, and the next iterations at 0.3s/it. The first iteration would largely skew the results, while this first iteration would be meaningless in a production setup where the model would only be traced during initialization.

Reply:
In https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat , it is written that `min()` should be used instead of average. 
I think the way it is implemented now is that it is run `self.args.repeat` times 10 times in a row and then ouf of those `self.args.repeat` times the min value is taken and average of the 10 times. 

I'm really not sure, whether it's better to time average or `min`. I had average in the beginning as well, but after reading  https://docs.python.org/2/library/timeit.html#timeit.Timer.repeat I switched to min. Since it is still averaged over 10 runs I guess `min()` is fine here as well, but I think you have more experience, so I let you decide :-) 

Regarding the second note, I think we should have a different timing mechanism for TPU. Maybe a simple `if` in the code is enough?

Reply:
Right, I guess that's okay. After reading a bit more on the subject, it seems that the choice really depends on the distribution so I don't have any particular insights. The average 10 times before minimum seems robust enough to me.

> Regarding the second note, I think we should have a different timing mechanism for TPU. Maybe a simple if in the code is enough?

Right, this is especially noticeable for TPUs, but actually really impacts the TorchScript performance as well. From what I'm seeing it doesn't seem you're actually leveraging torchscript/tracing the model anywhere, so it's probably not necessary yet. 

Maybe a `requires_tracing` flag for inference would be useful, where it does a few inferences at first before actually evaluating time spent, to make sure everything is correctly traced beforehand. Would be welcome in a second PR!